File: 3Rarm.py
import py_compile
import matplotlib.pyplot as plt
from matplotlib import style
import math

phideg=float(input("Enter the angle: "))
x=float(input("Enter the x coordinate: "))
y=float(input("Enter the y coordinate: "))
pi=3.141592653589793238462643383279502884
phi=phideg*pi/180
a1=1
a2=1
a3=1
wx=x-a3*math.cos(phi)
wy=y-a3*math.sin(phi)
#c2=(wx*wx+wy*wy-a1*a1-a2*a2)/(2*a1*a2)
#if c2<=1:   
''' s2_1=math.sqrt(1-c2*c2)
    s2_2=-(math.sqrt(1-c2*c2))
    theta2_1=math.atan2(s2_1,c2)
    theta2_2=math.atan2(s2_2,c2)
    denom_1=a1*a1+a2*a2+2*a1*a2*math.cos(theta2_1);
    denom_2=a1*a1+a2*a2+2*a1*a2*math.cos(theta2_2);
    s1_1=(wy*(a1+a2*math.cos(theta2_1))-a2*math.sin(theta2_1)*wx)/denom_1
    s1_2=(wy*(a1+a2*math.cos(theta2_2))-a2*math.sin(theta2_2)*wx)/denom_2
    c1_1=(wx*(a1+a2*math.cos(theta2_1))-a2*math.sin(theta2_1)*wy)/denom_1
    c1_2=(wx*(a1+a2*math.cos(theta2_2))-a2*math.sin(theta2_2)*wy)/denom_2
    theta1_1=math.atan2(s1_1,c1_1)
    theta1_2=math.atan2(s1_2,c1_2)
    theta3_1=phi-theta1_1-theta2_1
    theta3_2=phi-theta1_2-theta2_2 '''
if 1 :    
    alpha=math.atan(wy/wx)
    theta2_1=theta2_2=pi-math.acos((a1*a1+a2*a2-wx*wx-wy*wy)/(2*a1*a2))
    theta1_1=alpha-math.acos((wx*wx+wy*wy+a1*a1-a2*a2)/(2*a1*math.sqrt(wx*wx+wy*wy)))
    theta1_2=alpha+math.acos((wx*wx+wy*wy+a1*a1-a2*a2)/(2*a1*math.sqrt(wx*wx+wy*wy)))
    #print("%f %f \n%f %f \n%f %f"%(theta1_1*180/pi,theta1_2*180/pi,theta2_1*180/pi,theta2_2*180/pi,theta3_1*180/pi,theta3_2*180/pi))

    ax_1=a1*math.cos(theta1_1)
    ax_2=a1*math.cos(theta1_2)
    ay_1=a1*math.sin(theta1_1)
    ay_2=a1*math.sin(theta1_2)
    #bx_1=ax_1+a2*math.cos(theta1_1+theta2_1)
    #by_1=ay_1+a2*math.sin(theta1_1+theta2_1)
    #bx_2=ax_2+a2*math.cos(theta1_2+theta2_2)
    #by_2=ay_2+a2*math.sin(theta1_2+theta2_2)
    #bx=ax_1+a2*math.cos(theta1_1+theta2_1)
    #by=ay_1+a2*math.sin(theta1_1+theta2_1)
    bx=wx
    by=wy
    cx=x
    cy=y
    style.use('ggplot')
    #sol1_x=[0,ax_1,bx_1,cx]
    #sol1_y=[0,ay_1,by_1,cy]
    #sol2_x=[0,ax_2,bx_2,cx]
    #sol2_y=[0,ay_2,by_2,cy]
    sol1_x=[0,ax_1,bx,cx]
    sol1_y=[0,ay_1,by,cy]
    sol2_x=[0,ax_2,bx,cx]
    sol2_y=[0,ay_2,by,cy]
    plt.plot(range(5))
    plt.xlim(-3, 3)
    plt.ylim(-3, 3)
    plt.gca().set_aspect("equal", adjustable="box")
    plt.draw()
    plt.plot(sol1_x,sol1_y,'g',label='line one', linewidth=3)
    plt.plot(sol2_x,sol2_y,'c',label='line two',linewidth=3, linestyle='dashed')
    print(sol1_x ,sol1_y, sol2_x, sol2_y)
    plt.title('arm diagram')
    plt.ylabel('Y axis')
    plt.xlabel('X axis')
    plt.legend()
    plt.grid(True,color='k')
    plt.show()

else:
    print("Invalid Input")  
    

File: 3Rarm.cpp
#include<iostream>
#include<stdio.h>
#include<conio.h>
#include<math.h>
void polarxy(int link,float angle);
float link_x,link_y;

int main()
{   float phideg,x,y;
    link_x=0,link_y=0;
    std::cout<<"Enter the angle and x,y coordinates: \n";
    std::cin>>phideg>>x>>y;
    float const pi=3.141592653589793238462643383279502884;
    float phi=phideg*pi/180;
    int a1=1,a2=1,a3=1;
    float wx=x-a3*cos(phi);
    float wy=y-a3*sin(phi);
    float alpha=atan(wy/wx);
    float theta2_1=(pi-acos((a1*a1+a2*a2-wx*wx-wy*wy)/(2*a1*a2)));
    float theta2_2=(pi-acos((a1*a1+a2*a2-wx*wx-wy*wy)/(2*a1*a2)));
    float theta1_1=alpha-acos((wx*wx+wy*wy+a1*a1-a2*a2)/(2*a1*sqrt(wx*wx+wy*wy)));
    float theta1_2=alpha+acos((wx*wx+wy*wy+a1*a1-a2*a2)/(2*a1*sqrt(wx*wx+wy*wy)));
    //float theta2_1=alpha-theta1_1;
    //float theta2_2=alpha+theta1_2;
    if(sqrt(x*x+y*y)>(a1+a2+a3)|| ((a1*a1+a2*a2-wx*wx-wy*wy)/(2*a1*a2))>1)
    {
        theta1_1=atan(y/x);
        theta1_2=atan(y/x);
        theta2_1=atan(y/x);
        phi=atan(y/x);
    }
    std::cout<<theta1_1*180/pi<<"\t"<<theta2_1*180/pi<<"\t"<<phi*180/pi<<"\t";
    std::cout<<"\n The coordinates for link 1 is:\n";
    polarxy(a1,theta1_1);
    std::cout<<"\n The coordinates for link 2 is:\n";
    polarxy(a2,theta2_1);
    std::cout<<"\n The coordinates for link 3 is:\n";
    polarxy(a3,phi);
    

    return 0;
}

void polarxy(int link,float angle)
{
  link_x=link_x+link*cos(angle);
  link_y=link_y+link*sin(angle);
  std::cout<<link_x<<"\t"<<link_y<<"\t";

}